### PR TITLE
Polishing

### PR DESCRIPTION
**Pull request by jbduncan**
This PR is a follow-on to the discussion started on the bazel-dev mailing list at https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/bazel-dev/Q2owiR-e86s/ugrVUhn7AwAJ.

As per the commit message, this PR achieves the following:

```txt
- Use Java 8 idioms more consistently.
- Use newer Guava idioms more consistently.
- Apply some IntelliJ IDEA refactoring suggestions.
- Other changes made for readability and/or brevity.
```

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: src/main/java/com/google/devtools/build/lib/graph/DFS.java [hunk 1]
```diff
@@ -98,7 +95,8 @@ public void visit(Node<T> node, GraphVisitor<T> visitor) {
     Collection<Node<T>> edgeTargets = transpose
         ? node.getPredecessors() : node.getSuccessors();
     if (edgeOrder != null) {
-      List<Node<T>> mutableNodeList = Ordering.from(edgeOrder).sortedCopy(edgeTargets);
+      List<Node<T>> mutableNodeList =
```
Comment (laszlocsomor) [inline src/main/java/com/google/devtools/build/lib/graph/DFS.java]: What's the advantage of using streams instead of Ordering?

File: src/main/java/com/google/devtools/common/options/OptionsClassProvider.java [hunk 2]
```diff
@@ -19,6 +19,7 @@
  * A read-only interface for options parser results, which only allows to query the options of
  * a specific class, but not e.g. the residue any other information pertaining to the command line.
  */
+@FunctionalInterface
```
Comment (dslomov) [inline src/main/java/com/google/devtools/common/options/OptionsClassProvider.java]: Why this change? This is not a generic function interface.

Comment (jbduncan) [inline src/main/java/com/google/devtools/common/options/OptionsClassProvider.java]: @dslomov I made a reasonable guess that it was meant to effectively be a functional interface, because it currently has only one non-default method.

Is it likely for `OptionsClassProvider` to get new non-default methods in the future?

File: src/main/java/com/google/devtools/build/lib/graph/DFS.java (hunk 1 above)
Comment (jbduncan) [inline src/main/java/com/google/devtools/build/lib/graph/DFS.java]: @laszlocsomor The reason I decided to replace `Ordering` here with streams is because it seems that `Ordering` is now [obsolete for Java 8 users](https://google.github.io/guava/releases/22.0/api/docs/com/google/common/collect/Ordering.html) (for the most part).

If you think this change isn't worth it, please let me know and I'd be happy to revert it. :)

File: src/main/java/com/google/devtools/common/options/OptionsClassProvider.java (hunk 2 above)
Comment (dslomov) [inline src/main/java/com/google/devtools/common/options/OptionsClassProvider.java]: We are not consistently using `@FunctionalInterface` interface in the code, so let's not add it.

File: src/main/java/com/google/devtools/build/lib/graph/DFS.java (hunk 1 above)
Comment (laszlocsomor) [inline src/main/java/com/google/devtools/build/lib/graph/DFS.java]: Oh cool, I didn't know that, thanks!

File: src/main/java/com/google/devtools/common/options/OptionsClassProvider.java (hunk 2 above)
Comment (jbduncan) [inline src/main/java/com/google/devtools/common/options/OptionsClassProvider.java]: Okey dokey.

In that case, should the `@FunctionalInterface` lines that I added to other interfaces in this PR stay?

**Comment by jbduncan** [comment]
Thanks @laszlocsomor!

Do you have any thoughts on @dslomov's [comment](https://github.com/bazelbuild/bazel/pull/3462#discussion_r130846108) on my use of `@FunctionalInterface`? If not, then I'm happy to remove it from `OptionsClassProvider` as he requested.

**Comment by laszlocsomor** [comment]
You're welcome!

I'm fine either way, but I leave the decision up to @dslomov .

**Comment by jbduncan** [comment]
@dslomov I've just amended and force-pushed my last commit to no longer have the `@FunctionalInterface` on `OptionsClassProvider`, as you requested.

**Comment by dslomov** [comment]
Jenkins please test this

**Comment by dslomov** [comment]
test this please

**Comment by jbduncan** [comment]
@laszlocsomor @dslomov I have very little experience with Jenkins, so I'm struggling to navigate the UI and figure out why the ci.bazel.io build failed.

If you have any thoughts on how I could better navigate the Jenkins dashboard and interpret the logs, I would be really grateful!

**Comment by dslomov** [comment]
CI is it good shape, I'll merge (the failure is known, we are migrating to a new version of Jenkins and things are a bit rough this week)

**Comment by laszlocsomor** [comment]
I think your PR is fine. I see no sign of error, it's a mystery why Jenkins marked it as unstable.

We (specifically @damienmg) is in the middle of migrating to Jenkins Blue Ocean, and I suppose there are still some issues.

**Comment by jbduncan** [comment]
@dslomov @laszlocsomor Okay, great! I look forward to seeing this merged in when you get the time. :)

**Comment by jbduncan** [comment]
@dslomov @laszlocsomor Just wanted to follow up with you both. How are things doing with this PR? Is there anything I can do help move things forward?

**Comment by laszlocsomor** [comment]
Hey @jbduncan , sorry about the silence. @dslomov already imported this PR, it's approved, it should get submitted and pushed back to GitHub today.

**Comment by jbduncan** [comment]
That's great! Many thanks for your response @laszlocsomor, and thank you @dslomov very much for merging this in. :+1:

**Comment by laszlocsomor** [comment]
No problem! Thanks for your contribution!
